draft-ietf-mls-protocol.md

```diff
@@ -1148,39 +1166,42 @@ struct {
         case 1: T value;
     }
 } optional<T>;
-
-struct {
-    HPKEPublicKey public_key;
-    uint32 unmerged_leaves<0..2^32-1>;
-    opaque parent_hash<0..255>;
-} ParentNode;
 ~~~~~
 
-When computing the hash of a parent node, the `ParentNodeHashInput`
-structure is used:
+The tree hash of a leaf node is the hash of leaf's `LeafNodeHashInput` object which
+might include a Key Package depending on whether or not it is blank.
 
 ~~~~~
 struct {
     uint32 node_index;
-    optional<ParentNode> parent_node;
-    opaque left_hash<0..255>;
-    opaque right_hash<0..255>;
-} ParentNodeHashInput;
+    optional<KeyPackage> key_package;
+} LeafNodeHashInput;
 ~~~~~
 
-The `left_hash` and `right_hash` fields hold the hashes of the node's
-left and right children, respectively. When computing the hash of
-a leaf node, the hash of a `LeafNodeHashInput` object is used:
+Note that the `node_index` field contains the index of the leaf among the nodes
+in the tree, not its index among the leaves; `node_index = 2 * leaf_index`.
+
+Now the tree hash of any non-leaf node is recursively defined to be the hash of
+its `ParentNodeTreeHashInput`. This includes an optional `ParentNodeTreeHashData`
```

Comment:
```suggestion
its `ParentNodeTreeHashInput`. This includes an optional `ParentNode`
```

I prefer "ParentNode", since we use this for sending the tree as well as hashing.